functorch/functorch/experimental/cond.py

```diff
@@ -0,0 +1,143 @@
+    next_name = None
+    i = 0
+    while not next_name:
+        candidate = f"true_graph_{i}"
+        if hasattr(proxy_mode.tracer.root, candidate):
+            i += 1
+        else:
+            next_name = candidate
+
+    random_slug = ''.join(random.choices(string.digits, k=5))
+    true_name = next_name
+    false_name = f"false_graph_{i}"
+
+    proxy_mode.tracer.root.register_module(true_name, true_graph)
+    proxy_mode.tracer.root.register_module(false_name, false_graph)
+
+    with no_dispatch():
+        # This is not amazing.
+        # However, if we have nested operators that have a call_function
+        # in their graph that is not a torch op (ex: see conditional below, nested cond)
+        # we cannot get metadata for it from just looking at out vars.
+        # The reason is that an operation on the output of such an op is not
+        # evalauted as a torch.Tensor.
+        # So we execute the real true and false fn here and compare metadata
+        # inp_ops = [o for o in operands]
+        true_result = true_graph(operands)
+        false_result = false_graph(operands)
+        def recursive_compare_same(a, b):
+            assert(type(a) == type(b))
+            if isinstance(a, torch.Tensor):
+                assert(a.dtype == b.dtype)
+                assert(a.size() == b.size())
+                assert(a.stride() == b.stride())
+                assert(a.device == b.device)
+            elif isinstance(a, (list, tuple)):
+                assert(len(a) == len(b))
+                for i in range(0, len(a)):
+                    recursive_compare_same(a[i], b[i])
+
+        recursive_compare_same(true_result, false_result)
+
+    args = (pred, true_graph, false_graph, operands)
+
+    proxy_args = pytree.tree_map(_unwrap_proxy, args)
+
+    proxy_res = proxy_mode.tracer.create_proxy('call_function', func_overload, proxy_args, kwargs,
+                                            name="conditional")
+    return proxy_res
+
+
+def cond_dense(pred, true_fn, false_fn, *operands):
+    mode = torch._C._get_torch_dispatch_mode()
+    assert (mode is None), "Mode should never be enabled for CPU key"
+    if pred:
+        return true_fn(*operands)
+    else:
+        return false_fn(*operands)
+
+
+def cond_autograd(pred, true_fn, false_fn, *operands):
+    # TODO: support autograd
+    flat_operands, _ = tree_flatten([true_fn, false_fn] + [operands])
+    assert all([not f.requires_grad for f in flat_operands
+                if isinstance(f, torch.Tensor)])
+
+    guard = ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.AutogradCPU))
+    return cond(pred, true_fn, false_fn, *operands)
+
+
+def python_fallback(op):
+    def inner(*args, **kwargs):
+        mode = torch._C._get_torch_dispatch_mode()
+        assert (mode is not None), "Mode should always be enabled for python fallback key"
+        with suspend_mode(mode):
+            res = trace_cond(mode, op, args, kwargs)
+        return res
```

Comment:
This registration is not quite right. To register to the Python key implies that the implementation works for *any* Python mode and for *any* tensor subclass with `__torch_dispatch__` on it. To actually make good on this promise, you must actually call the `__torch_dispatch__` method that was provided in this way.

What this implementation does, instead, is call into a hard-coded implementation of the mode that is suitable for ProxyTensor. This is fine, but the python fallback mustn't imply that it can be used in other situations, which is what it will blindly crash through here. So you should at least assert that the mode stack has exactly one mode in it and it is proxy tensor. Which, speaking of which, means that fake tensor mode is not going to work properly in that case...

The easiest fix may just to be to write the generalized version of this logic. That just means recreating the python fallback key in C++ in PythonFallbackKernel.cpp; but you get to pass functions in the args instead. Then, ProxyTensorMode would be responsible for seeing that cond is being called and handling it properly. You could put your trace_cond implementation directly in ProxyMode, or you could add a little registration system for custom trace handling and then register your trace cond that way.

Reply:
I like the idea of a registration system. A much earlier version of this PR had this going through `__torch_dispatch__` on the mode, which in turn would call back to this in one version, and had the logic inside Proxy in another. 

> That just means recreating the python fallback key in C++ in PythonFallbackKernel.cpp; but you get to pass functions in the args instead.

What does moving it to C++ give us? Just sharing more of the stack with how we dispatch today?

Reply:
Ah, yeah, we absolutely should.